warehouse/transformer/transformer.go

```diff
@@ -71,22 +72,47 @@ func (t *Transformer) Transform(_ context.Context, clientEvents []types.Transfor
 		t.statsFactory.NewTaggedStat("warehouse_dest_transform_output_failed_events", stats.HistogramType, tags).Observe(float64(len(res.FailedEvents)))
 	}()
 
-	for _, clientEvent := range clientEvents {
-		r, err := t.processWarehouseMessage(c, &clientEvent)
-		if err != nil {
-			res.FailedEvents = append(res.FailedEvents, transformerResponseFromErr(&clientEvent, err))
-			continue
+	results := make(chan types.TransformerResponse, len(clientEvents))
+	done := make(chan struct{})
+
+	g := errgroup.Group{}
+	g.SetLimit(t.config.concurrentTransformations.Load())
+
+	go func() {
+		defer close(done)
+		for r := range results {
+			if r.Error != "" {
+				res.FailedEvents = append(res.FailedEvents, r)
+			} else {
+				res.Events = append(res.Events, r)
+			}
 		}
+	}()
 
-		res.Events = append(res.Events, lo.Map(r, func(item map[string]any, index int) types.TransformerResponse {
-			clientEvent.Metadata.SourceDefinitionType = "" // TODO: Currently, it's getting ignored during JSON marshalling Remove this once we start using it.
-			return types.TransformerResponse{
-				Output:     item,
-				Metadata:   clientEvent.Metadata,
-				StatusCode: http.StatusOK,
+	for i := range clientEvents {
+		event := clientEvents[i]
+		event.Metadata.SourceDefinitionType = "" // TODO: Currently, it's getting ignored during JSON marshalling Remove this once we start using it.
+
+		g.Go(func() error {
+			r, err := t.processWarehouseMessage(c, &event)
+			if err != nil {
+				results <- transformerResponseFromErr(&event, err)
+				return nil
+			}
+			for _, item := range r {
```

Comment:
Why do we have a 2nd routine for populating `res`? Is it to avoid concurrent access to the slice? If yes, can we use a lock instead?